Sources/HummingbirdHTTP2/HTTP2ServerConnectionManager+StateMachine.swift

```diff
@@ -0,0 +1,286 @@
+//
+// Copyright (c) 2024 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import NIOCore
+import NIOHTTP2
+
+extension HTTP2ServerConnectionManager {
+    struct StateMachine {
+        var state: State
+
+        init() {
+            self.state = .active(.init())
+        }
+
+        mutating func streamOpened(_ id: HTTP2StreamID) {
+            switch self.state {
+            case .active(var activeState):
+                activeState.openStreams.insert(id)
+                activeState.lastStreamId = id
+                self.state = .active(activeState)
+
+            case .closing(var closingState):
+                closingState.openStreams.insert(id)
+                closingState.lastStreamId = id
+                self.state = .closing(closingState)
+
+            case .closed:
+                break
+            }
+        }
+
+        enum StreamClosedResult {
+            case startIdleTimer
+            case close
+            case none
+        }
+
+        mutating func streamClosed(_ id: HTTP2StreamID) -> StreamClosedResult {
+            switch self.state {
+            case .active(var activeState):
+                activeState.openStreams.remove(id)
+                self.state = .active(activeState)
+                if activeState.openStreams.isEmpty {
+                    return .startIdleTimer
+                } else {
+                    return .none
+                }
+
+            case .closing(var closingState):
+                closingState.openStreams.remove(id)
+                if closingState.openStreams.isEmpty, closingState.sentSecondGoAway == true {
+                    self.state = .closed
+                    return .close
+                } else {
+                    self.state = .closing(closingState)
+                    return .none
+                }
+
+            case .closed:
+                return .none
+            }
+        }
+
+        enum TriggerGracefulShutdownResult {
+            case sendGoAway(pingData: HTTP2PingData)
+            case none
+        }
+
+        mutating func triggerGracefulShutdown() -> TriggerGracefulShutdownResult {
+            switch self.state {
+            case .active(let activeState):
+                let closingState = State.ClosingState(from: activeState)
+                self.state = .closing(closingState)
+                return .sendGoAway(pingData: closingState.goAwayPingData)
+
+            case .closing:
+                return .none
+
+            case .closed:
+                return .none
+            }
+        }
+
+        enum ReceivedPingResult {
+            case sendPingAck(pingData: HTTP2PingData)
+            case enhanceYouCalmAndClose(lastStreamId: HTTP2StreamID)
```

Comment:
What does this mean?

Reply:
Enhance you calm is an unofficial status code Twitter defined for when a client sends too many requests (In this case too many pings). I've added a comment

Reply:
I believe twitter actually spells it "Enhance your calm", not "Enhance you calm"

Reply:
🤣 that's a typo